### PR TITLE
A0-1409 Validator Network handshake

### PR DESCRIPTION
**Pull request by lesniak43**
# Description

Implementing proper handshake for validator network.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have added tests
- I have created new documentation

### CODE REVIEW

**Review by timorl — APPROVED** [review]
Najs.

File: finality-aleph/src/validator_network/handshake.rs [hunk 1]
```diff
@@ -47,6 +52,37 @@ impl From<ReceiveError> for HandshakeError {
...
+
+    /// Verify the Response sent by the peer.
+    fn verify(&self, peer_id: &AuthorityId, challenge: &Challenge) -> bool {
+        verify(peer_id, &challenge.encode(), &self.0)
+    }
+}
+
 /// Performs the handshake. The goal is to obtain ID of the peer,
 /// and split the communication stream into two halves.
 /// Current version makes an unrealistic assumption that the peer is not malicious.
```
Comment (pmikolajczyk41) [inline finality-aleph/src/validator_network/handshake.rs:88]: that's not true anymore, is it?

File: finality-aleph/src/validator_network/handshake.rs [hunk 2]
```diff
@@ -76,15 +125,107 @@ pub async fn v0_handshake<S: Splittable>(
...
+
+    fn assert_send_error(e: HandshakeError) {
+        match e {
+            HandshakeError::SendError(_) => (),
+            x => panic!(
+                "should end with HandshakeError::SendError, but we got {:?}",
+                x
+            ),
+        };
+    }
```
Comment (pmikolajczyk41) [inline finality-aleph/src/validator_network/handshake.rs:215]: `assert_matches!`?

**Review by pmikolajczyk41 — APPROVED** [review]
cool, clean code :+1:

File: finality-aleph/src/validator_network/handshake.rs (hunk 2 above)
Comment (lesniak43) [reply finality-aleph/src/validator_network/handshake.rs:215]: unstable :(

File: finality-aleph/src/validator_network/handshake.rs (hunk 1 above)
Comment (lesniak43) [reply finality-aleph/src/validator_network/handshake.rs:88]: true :)

**Pull request merged**